File: pi-euler-coprime-algorithm.swift
//: ## Computing Pi as the Probability of any two numbers being relatively prime (i.e., coprime)
//: ### Uses Euclid's Algorithm
import UIKit
let equation = #imageLiteral(resourceName: "Screen Shot 2021-03-17 at 2.06.05 PM.png")

// Wrapper for recursive GCD function to find Greatest Common Divisor of two integers
func gcd(a: Int, b: Int) -> Int {
    
    // Recursive GCD
    /// Uses Euler's method
    /// - Parameters:
    ///   - a: The first number
    ///   - b: The second number
    /// - Returns: The greatest common divisor
    func gcdRecursive(a: Int, b: Int) -> Int {
        return b == 0 ? a : gcdRecursive(a: b, b: a % b)
    }
    
    var a1 = abs(a)
    var b1 = abs(b)
    if (b1 > a1) { swap(&a1, &b1) }
    
    return gcdRecursive(a: a1, b: b1)
    
}


// Compute pi Usiing Euler's Method
// Where copprimes/total = 6 / pi^2

var randoms = [Int]()
for _ in 0..<50000 {
    randoms.append(Int.random(in: 1..<250000))
}

var coprimeCount = 0
var totalCount   = 0
var index        = 0
var piSoFar      = 0.0

for _ in 0..<randoms.count / 2 {
    let a = randoms[index]
    let b = randoms[index + 1]
    if gcd(a: a, b: b) == 1 {
        coprimeCount += 1
    }
    piSoFar = sqrt(6 * Double(totalCount) / Double(coprimeCount))
    totalCount += 1
    index += 2
}

let piFinal = piSoFar

File: pi-euler-coprime-algorithm.py
# By Michael Stebel using Euler's method and the Riemann Zeta Function
import random
import math
from functools import cache

@cache

# Recursive greatest common divisor function
def gcd(a: int, b: int) -> int:
    if b == 0:
        return a
    else:
        return gcd(b, a % b)

# Create an array of integers and fill it with n random integers
# Lots of iterations are needed to converge to pi!
n = 100000000
randoms = [int]
for i in range(n):
    rnd_number = random.randint(1, n)
    randoms.append(rnd_number)

# The probability that two random numbers are co-prime converges to 6/pi^2 as n -> infinity
total_count: int = 0
coprimes: int = 1
index: int = 1
pi_so_far: float = 0.0
count = int(len(randoms) / 2)
a2: int = 0
b2: int = 0

for i in range(count):
    a2 = randoms[index]
    b2 = randoms[index + 1]
    # Swap if b2 is larger than a2
    if b2 > a2:
        [a2, b2] = [b2, a2]
    if gcd(a2, b2) == 1:
        coprimes += 1
    total_count += 1
    index += 2
    pi_so_far = math.sqrt(6.0 * total_count / coprimes)
    print(pi_so_far)
